packages/back-end/src/services/experimentNotifications.ts

```diff
@@ -0,0 +1,205 @@
+import { ExperimentWarningNotificationPayload } from "../types/ExperimentNotification";
+import { IfEqual } from "../util/types";
+
+// This ensures that the two types remain equal.
+
+// TODO: extend with experiment info
+type ExperimentNotificationFromCode = ExperimentWarningNotificationPayload["type"];
+
+type ExperimentWarningNotificationData = IfEqual<
+  ExperimentNotificationFromCode,
+  ExperimentNotification,
+  ExperimentWarningNotificationPayload,
+  never
+>;
+
+const dispatchEvent = async (
+  context: Context,
+  data: ExperimentWarningNotificationData
+) => {
+  const payload: ExperimentWarningNotificationEvent = {
+    event: "experiment.warning",
+    object: "experiment",
+    data,
+    user: {
+      type: "dashboard",
+      id: context.userId,
+      email: context.email,
+      name: context.userName,
+    },
+  };
+
+  const emittedEvent = await createEvent(context.org.id, payload);
+
+  if (!emittedEvent) throw new Error("Error while creating event!");
+
+  new EventNotifier(emittedEvent.id).perform();
+};
+
+export const memoizeNotification = async ({
+  context,
+  experiment,
+  type,
+  triggered,
+  dispatch,
+}: {
+  context: Context;
+  experiment: ExperimentInterface;
+  type: ExperimentNotification;
+  triggered: boolean;
+  dispatch: () => Promise<void>;
+}) => {
+  if (triggered && experiment.pastNotifications?.includes(type)) return;
+  if (!triggered && !experiment.pastNotifications?.includes(type)) return;
+
+  await dispatch();
+
+  const pastNotifications = triggered
+    ? [...(experiment.pastNotifications || []), type]
+    : (experiment.pastNotifications || []).filter((t) => t !== type);
+
+  await updateExperiment({
+    experiment,
+    context,
+    changes: {
+      pastNotifications,
+    },
+  });
+};
+
+export const notifyAutoUpdate = ({
+  context,
+  experiment,
+  success,
+}: {
+  context: Context;
+  experiment: ExperimentInterface;
+  success: boolean;
+}) =>
+  memoizeNotification({
+    context,
+    experiment,
+    type: "auto-update",
+    triggered: !success,
+    dispatch: () =>
+      dispatchEvent(context, {
+        type: "auto-update",
+        success,
+        experimentId: experiment.id,
+        experimentName: experiment.name,
+      }),
+  });
+
+export const MINIMUM_MULTIPLE_EXPOSURES_PERCENT = 0.01;
+
+const notifyMultipleExposures = async ({
+  context,
+  experiment,
+  results,
+  snapshot,
+}: {
+  context: Context;
+  experiment: ExperimentInterface;
+  results: ExperimentReportResultDimension;
+  snapshot: ExperimentSnapshotDocument;
+}) => {
+  const totalsUsers = results.variations.reduce(
+    (totalUsersCount, { users }) => totalUsersCount + users,
+    0
+  );
+  const percent = snapshot.multipleExposures / totalsUsers;
+  const multipleExposureMinPercent =
+    context.org.settings?.multipleExposureMinPercent ??
+    MINIMUM_MULTIPLE_EXPOSURES_PERCENT;
+
+  const triggered = multipleExposureMinPercent <= snapshot.multipleExposures;
```

Comment:
Ah shoot. I missed this. This needs to be the following, no?

```
  const triggered = multipleExposureMinPercent <= percent;
```

I guess this is one of the advantages of using the existing code first :)

Reply:
You can see this logic (or rather `multipleExposureMinPercent > percent`) in `multiExposureCheck`

Reply:
Fixed!